src/main/kotlin/no/digdir/fdk/parseservice/kafka/KafkaReasonedEventCircuitBreaker.kt

```diff
@@ -0,0 +1,81 @@
+package no.digdir.fdk.parseservice.kafka
+
+import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker
+import io.micrometer.core.instrument.Metrics
+import no.digdir.fdk.parseservice.model.RecoverableParseException
+import no.digdir.fdk.parseservice.model.UnrecoverableParseException
+import no.digdir.fdk.parseservice.handler.DatasetHandler
+import no.fdk.dataset.DatasetEvent
+import no.fdk.dataset.DatasetEventType
+import no.fdk.rdf.parse.RdfParseEvent
+import no.fdk.rdf.parse.RdfParseResourceType
+import org.apache.avro.specific.SpecificRecord
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+import kotlin.time.measureTimedValue
+import kotlin.time.toJavaDuration
+
+
+@Component
+open class KafkaReasonedEventCircuitBreaker(
+    private val producer: KafkaRdfParseEventProducer,
+    private val datasetHandler: DatasetHandler
+) {
+    @CircuitBreaker(name = "rdf-parse")
+    open fun process(record: ConsumerRecord<String, SpecificRecord>) {
+        LOGGER.debug("Received message - offset: " + record.offset())
+        val event = record.value()
+
+        val resourceType = when (event) {
+            is DatasetEvent -> RdfParseResourceType.DATASET
+            else -> throw UnrecoverableParseException("Unknown event type")
+        }
+
+        try {
+            event.let {
+                if (it.type == DatasetEventType.DATASET_REASONED) {
+                    parseAndProduce(it.fdkId.toString(), it.graph.toString(), it.timestamp, resourceType)
+                }
+            }
+        } catch (e: RecoverableParseException) {
+            LOGGER.warn("Recoverable parsing error: " + e.message)
+            Metrics.counter(
+                "rdf_parse_error",
+                "type", resourceType.toString().lowercase()
+            ).increment()
+            throw e
+        } catch (e: UnrecoverableParseException) {
+            LOGGER.error("Unrecoverable parsing error: " + e.message)
+            Metrics.counter(
+                "rdf_parse_error",
+                "type", resourceType.toString().lowercase()
+            ).increment()
+            throw e
+        }
+    }
+
+    private fun parseAndProduce(fdkId: String, graph: String, timestamp: Long, type: RdfParseResourceType) {
+        val timeElapsed = measureTimedValue {
+            LOGGER.debug("Parse dataset - id: $fdkId")
+            val json = when (type) {
+                RdfParseResourceType.CONCEPT -> LOGGER.error("Parse of concepts not implemented")
```

Comment:
warning isteden for error

Reply:
c8bf881aab98eff96f0fbdc210101ac347a3600c